gpflow/base.py

```diff
@@ -138,10 +138,6 @@ def transform(self, new_transform):
     def trainable(self):
         return self._unconstrained.trainable
 
-    @trainable.setter
```

Comment:
Would be possible to throw a more descriptive error message when a user tries to set `trainable`? Something that points them to using `gpflow.utilities.set_trainable` for example.

Reply:
hmmm ... not sure. the only way I can think of doing that is to put the setter back and raise an error if it's used, but having the setter there just so it can raise an error sounds like a source for confusion. how about I document it in the getter?

Reply:
A user will not be able to set a trainable attribute without an error: it is a property. I would like to have a clear message in the docs rather than a setter with a custom error. The exception will be too much honestly.

Reply:
i'll add it to the docs

Reply:
In this case I think a custom setter that raises an error message would significantly help user experience. Otherwise it is kind of confusing - it has a `trainable` attribute like a tf.Variable, but there it is simply settable, the whole point of removing it was to make it *easier* for users not to do it wrong - we shouldn't make it harder for them.

Reply:
what do you mean by it's settable on `tf.Variable`
```python3
>>> v = tf.Variable(1)
>>> v.trainable
True
>>> v.trainable = False
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
AttributeError: can't set attribute
```

Reply:
You're right, sorry my bad... should've tried it out myself first. I've tried looking for tensorflow documentation but could not find any mention of how they would suggest you handle wanting to change the trainable status of a variable! This makes me wonder if we'll run into some subtle bugs with tf.function() later ? does `set_trainable()` need a "cannot run inside tf.function / only supports eager mode" check?

Reply:
good catch. I could add a warning that prints every time `set_trainable` is used inside `tf.function`

man, i'm yet to be convinced by this `tf.function` thing. I reckon you basically need to assume everything that's not a tensor is immutable inside a `tf.function`. so much room for error

Reply:
TensorFlow never allowed to change trainable status. Although, there are ways of doing it. No setter gives the same behaviour as a `tf.Variable`. That was a reason why _I said before that removing is okay_.

The `tf.function` is not affected by a `trainable` status of the `tf.Variable`, although you may find cases where it can be a case.

Reply:
@awav i don't understand. do you mind rephrasing?

Reply:
I'm happy to remove the setter outright. What I was concerned about is calling set_trainable() inside a function that will then get wrapped in tf.function(), and this leading to behaviour that perhaps wasn't expected by the caller (because set_trainable() would probably only be called during the tracing step, not each time)... so a warning when executing in not-eager-mode might be helpful.

Reply:
@st-- i've added a warning that shows only if you use `set_trainable` inside `tf.function`

Reply:
and now removed it